ee/billing/quota_limiting.py

```diff
@@ -87,42 +101,174 @@ def org_quota_limited_until(organization: Organization, resource: QuotaResource)
+            return {
+                "quota_limited_until": None,
+                "quota_limiting_suspended_until": quota_limiting_suspended_until,
+                "needs_save": needs_save,
+            }
+    elif trust_score == 10:
+        # If limitng suspended was set in the previous period or was never set, update it.
+        if (
+            not quota_limiting_suspended_until
+            or (
+                datetime.fromtimestamp(quota_limiting_suspended_until)
+                - timedelta(QUOTA_LIMIT_MEDIUM_HIGH_TRUST_GRACE_PERIOD_DAYS)
+            ).timestamp()
+            < billing_period_start
+        ):
+            # Medium high trust, retain data for three days
+            report_organization_action(
+                organization, "quota limiting suspended", properties={"current_usage": usage + todays_usage}
+            )
+            quota_limiting_suspended_until = round(
+                (today_end + timedelta(days=QUOTA_LIMIT_MEDIUM_HIGH_TRUST_GRACE_PERIOD_DAYS)).timestamp()
+            )
+            needs_save = True
+            summary["quota_limiting_suspended_until"] = quota_limiting_suspended_until
+            return {
+                "quota_limited_until": None,
+                "quota_limiting_suspended_until": quota_limiting_suspended_until,
+                "needs_save": needs_save,
+            }
+        elif today_end.timestamp() <= quota_limiting_suspended_until:
+            # Return existing quota limiting date
+            return {
+                "quota_limited_until": None,
+                "quota_limiting_suspended_until": quota_limiting_suspended_until,
+                "needs_save": needs_save,
+            }
+    return {
+        "quota_limited_until": billing_period_end,
+        "quota_limiting_suspended_until": None,
+        "needs_save": needs_save,
+    }
 
 
 def sync_org_quota_limits(organization: Organization):
+    _, today_end = get_current_day()
     if not organization.usage:
         return None
 
     for resource in [QuotaResource.EVENTS, QuotaResource.RECORDINGS, QuotaResource.ROWS_SYNCED]:
+        previously_quota_limited_team_tokens = list_limited_team_attributes(
+            resource, QuotaLimitingCaches.QUOTA_LIMITER_CACHE_KEY
+        )
         team_attributes = get_team_attribute_by_quota_resource(organization, resource)
-        quota_limited_until = org_quota_limited_until(organization, resource)
-
-        if quota_limited_until:
-            add_limited_team_tokens(resource, {x: quota_limited_until for x in team_attributes})
+        result = org_quota_limited_until(organization, resource, previously_quota_limited_team_tokens)
+
+        if result:
+            quota_limited_until = result.get("quota_limited_until")
+            limiting_suspended_until = result.get("quota_limiting_suspended_until")
+            needs_save = result.get("needs_save")
+            if needs_save:
+                organization.save(update_fields=["usage"])
```

Comment:
I was trying to minimize the number of functions in which we call `.save()` ... Maybe it will come in handy in the future, maybe it won't, I'm happy to move `.save()` into `org_quota_limited_until`.